### ISSUE TITLE
loss.backward() hangs after burn update 0.12 -> 0.13

### ISSUE THREAD

**Issue by kurnevsky** [tracel-ai/burn#1702]
**Describe the bug**
`loss.backward()` started to hang in my tests for both ndarray and wgpu.

**To Reproduce**
1. `git clone https://github.com/pointsgame/oppai-rs.git -b burn && cd oppai-rs/zero-burn`
2. `cargo test --release -- train_ndarray`
3. It hangs in [this line](https://github.com/pointsgame/oppai-rs/blob/d967bd0cc881afec108c9c2878c35fb003098e4c/zero-burn/src/model.rs#L241).

**Desktop:**
 - OS: linux

**Additional context**
I haven't tried to extract it to a minimal example, but the test is pretty straightforward. Say if you need it.

**Comment by nathanielsimard** [comment]
@kurnevsky Found the problem, working on it, thanks !

**Comment by louisfd** [comment]
@kurnevsky I fixed my recursive code in #1710 , ran the test in your project and it's fast again. I guess we'll have to release a new patch soon 😅

**Comment by kurnevsky** [comment]
Thanks! Checked the MR - it's possible to use recursive approach as well, you just need to pass `visited` nodes as an additional argument. But I have no idea what's going on there, so up to you how it should be :)
